### PR TITLE
feat(schema): Inference controls and improvements

### PR DESCRIPTION
**Pull request by freiksenet**
# Evolution of schema customization

## tl/dr

Try it now with `@schema-customization` tag.

```
yarn add gatsby@schema-customization
```

Use resolver directives/extensions to add default arguments or/and resolvers to fields.

```graphql
type MyType @infer {
  date: Date @dateformat(formatString: "DD MMM", locale: "fi")
  image: File @fileByRelativePath
  authorByEmail: Author @link(by: "email")
}
```

Use `@dontInfer` and enjoy better performance with no inferrence.

As usual, report issues in #12272

## Summary

After about a month of testing schema customization both here and in pre-release we determined a couple of issues. We set up to do this because we wanted to remove uncertainty in people's schemas, when the data changes. However, the original design allowed some uncertainties anyway. In addition, we have made inferrence a more heavy process, trading performance for consistency and didn't really provide a way to opt out of it completely. To summarize:

* Resolvers and arguments of fields like Date and File was determined by inferred data
* There was no easy way to use arguments/resolvers to override the above
* Inferrence was run even when `@dontInfer` flag was on
* There was no way to control inference outside of SDL

Therefore we have some proposed changes to the way we do inferrence and some deprecations.

## noDefaultResolvers and inferrence modes

First of all, we are deprecating `noDefaultResolvers` argument of `infer` and `dontInfer`. We feel it was confusing and very catch all and didn't even add resolvers at the end. We will support `noDefaultResolvers` until version 3, after which `@infer` behaviour will become a default and `noDefaultResolvers` will be removed.

We didn't want to break things, so we keep old default behaviour, even though we think it's not optimal. Add explicit `@infer` and resolver extensions to fields to be future proof.

### Default (deprecated in v3)

Applies with no `@infer` or there is `@infer(noDefaultResolvers: false)` on type.

Type gets all inferred fields added. If type has defined fields of types `Date`, `File` and any other node, and we inferred that they should have resolvers/args, resolvers/args will be added to type with a warning.

### Strict inference (future default in v3)

Applies with `@infer` or `@infer(noDefaultResolvers: true)`.

Type gets all inferred fields added. Existing fields won't automatically get resolvers

### No inferrence 

Applies with `@dontInfer` or `@dontInfer(noDefaultResolvers: true)`.

Inferrence won't run at all. Existing fields won't automatically get resolvers

### No new fields with default resolvers (deprecated, removed in v3)

Applies with `@dontInfer(noDefaultResolvers: false)`

Inferrence will run, but fields won't be added.  If type has defined fields of types `Date`, `File` and any other node, and we inferred that they should have resolvers/args, resolvers/args will be added to type with a warning.

## Resolver extensions 

To add an explicit workraround to not getting resolvers when there is no inferred ones, one can now use explicit resolver directives.

```graphql
type MyType @infer {
  date: Date @dateformat(formatString: "DD MMM", locale: "fi")
  image: File @fileByRelativePath
  authorByEmail: Author @link(by: "email")
}
```

## Type Builders and extensions

You can now apply configuration to type builder types through extension property on them.

```js
schema.createObjectType({
  name: MyType,
  extensions: {
    infer: true,
  },
  fields: {
    date: {
      type: "Date",
      extensions: {
        dateformat: {
          formatString: "DD MMM", locale: "fi"
        }
      }
    }
  },
})
```

### CODE REVIEW

File: packages/gatsby/src/schema/__tests__/queries.js [hunk 1]
```diff
@@ -45,7 +45,7 @@ describe(`Query schema`, () => {
       } else if (api === `createResolvers`) {
         return [
           args[0].createResolvers({
-            Frontmatter: {
+            MarkdownFrontmatter: {
```
Comment (stefanprobst) [inline packages/gatsby/src/schema/__tests__/queries.js]: Does the custom name for nested fields still work as well?

File: packages/gatsby/src/schema/add-field-resolvers.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  }
+  if (!field.args || _.isEmpty(field.args)) {
+    fieldConfig.args = {
+      ...dateResolver.args,
+    }
+    if (defaultFormat) {
+      fieldConfig.args.formatString.defaultValue = defaultFormat
+    }
+    if (defaultLocale) {
+      fieldConfig.args.formatString.defaultLocale = defaultLocale
```
Comment (stefanprobst) [inline packages/gatsby/src/schema/add-field-resolvers.js]:
```suggestion
      fieldConfig.args.formatString.defaultValue = defaultLocale
```

File: packages/gatsby/src/schema/add-field-resolvers.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        case `date`: {
+          addDateResolver({
+            typeComposer,
+            fieldName,
+            options,
+          })
+          break
+        }
+        case `link`: {
+          typeComposer.extendField(fieldName, {
```
Comment (stefanprobst) [inline packages/gatsby/src/schema/add-field-resolvers.js]: Just to quickly confirm: we only allow one extension per field, so it's ok to `extendField`, otherwise we'd have to wrap the original resolver.

File: packages/gatsby/src/schema/infer/add-inferred-fields.js [hunk 4]
```diff
@@ -84,66 +66,28 @@ const addInferredFieldsImpl = ({
...
+    ;({ key, fieldConfig } = getFieldConfig({
+      ...selectedField,
+      schemaComposer,
+      typeComposer,
+      nodeStore,
+      prefix,
+      typeMapping,
+    }))
+
     let namedInferredType = fieldConfig.type
```
Comment (stefanprobst) [inline packages/gatsby/src/schema/infer/add-inferred-fields.js]: Is this still used?

File: packages/gatsby/src/schema/infer/add-inferred-fields.js [hunk 5]
```diff
@@ -293,18 +231,26 @@ const getFieldConfigFromFieldNameConvention = ({
   // (ii) hinders reusing types.
   if (linkedTypes.length > 1) {
     const typeName = linkedTypes.sort().join(``) + `Union`
-    type = schemaComposer.getOrCreateUTC(typeName, utc => {
-      const types = linkedTypes.map(typeName =>
-        schemaComposer.getOrCreateOTC(typeName)
-      )
-      utc.setTypes(types)
-      utc.setResolveType(node => node.internal.type)
+    type = UnionTypeComposer.createTemp({
```
Comment (stefanprobst) [inline packages/gatsby/src/schema/infer/add-inferred-fields.js]: Do we still have to use a temporary composer now that we choose from `possibleFields` earlier?

File: packages/gatsby/src/schema/infer/add-inferred-fields.js [hunk 6]
```diff
@@ -293,18 +231,26 @@ const getFieldConfigFromFieldNameConvention = ({
...
     })
+    type.setResolveType(node => node.internal.type)
   } else {
     type = linkedTypes[0]
   }
 
-  return { type, resolve: link({ by: foreignKey || `id` }) }
+  return {
+    type,
+    extensions: {
```
Comment (stefanprobst) [inline packages/gatsby/src/schema/infer/add-inferred-fields.js]: We should also use the extension in `getFieldConfigFromMapping`

File: packages/gatsby/src/schema/__tests__/queries.js (hunk 1 above)
Comment (freiksenet) [reply packages/gatsby/src/schema/__tests__/queries.js]: Yeah, that is artifact from old test changes when it didn't. thanks

File: packages/gatsby/src/schema/infer/add-inferred-fields.js (hunk 5 above)
Comment (freiksenet) [reply packages/gatsby/src/schema/infer/add-inferred-fields.js]: Everything in add-inferred-fields is a temp composer, it's only added in merge.

File: packages/gatsby/src/schema/infer/add-inferred-fields.js (hunk 6 above)
Comment (freiksenet) [reply packages/gatsby/src/schema/infer/add-inferred-fields.js]: thanks

File: packages/gatsby/src/schema/add-field-resolvers.js (hunk 3 above)
Comment (freiksenet) [reply packages/gatsby/src/schema/add-field-resolvers.js]: Yeah!

File: packages/gatsby/src/schema/infer/add-inferred-fields.js (hunk 4 above)
Comment (freiksenet) [reply packages/gatsby/src/schema/infer/add-inferred-fields.js]: good catch

File: packages/gatsby/src/schema/infer/index.js [hunk 7]
```diff
@@ -57,36 +20,28 @@ const addInferredTypes = ({
...
+  const nodeTypesToInfer = []
+
   typeNames.forEach(typeName => {
     let typeComposer
-    let inferConfig
     if (schemaComposer.has(typeName)) {
       typeComposer = schemaComposer.getOTC(typeName)
-      inferConfig = getInferConfig(typeComposer)
-      if (inferConfig.infer) {
+      // Infer we have enable infer or if it's "@dontInfer" but we have "addDefaultResolvers: false"
```
Comment (stefanprobst) [inline packages/gatsby/src/schema/infer/index.js]: should it be `addDefaultResolvers: true` in the comment?

Comment (freiksenet) [reply packages/gatsby/src/schema/infer/index.js]: yes

File: packages/gatsby/src/schema/infer/add-inferred-fields.js (hunk 5 above)
Comment (stefanprobst) [reply packages/gatsby/src/schema/infer/add-inferred-fields.js]: ah cool, just asking in general if we still need to work with temporary composers (instead of directly mutating)? the original motivation for temp composers was because we inferred for all fields even if they had conflicting field names, and only then decided which one should have priority. now that we decide before inferring do we still have to do it? could remove a bit of complexity, e.g. [here](https://github.com/gatsbyjs/gatsby/blob/774ada430f6dc67378facc6e2493976740f84f4f/packages/gatsby/src/schema/infer/add-inferred-fields.js#L297-L317). if we still need to do it, maybe a short comment in `addInferredType`?

File: packages/gatsby/src/schema/infer/index.js [hunk 8]
```diff
@@ -1,49 +1,12 @@
...
-    exampleValue,
-    inferConfig: getInferConfig(typeComposer),
-    typeMapping,
-    parentSpan,
-  })
-  return typeComposer
-}
+const { mergeInferredComposer } = require(`./merge-inferred-composer`)
 
 const addInferredTypes = ({
```
Comment (stefanprobst) [inline packages/gatsby/src/schema/infer/index.js]: Should we rename `addInferredTypes`, now that they don't really add anything? `getNodeTypesToInfer` is a bit clunky maybe but something like that?

Comment (freiksenet) [reply packages/gatsby/src/schema/infer/index.js]: Good idea.

Comment (freiksenet) [reply packages/gatsby/src/schema/infer/index.js]: Oh, they do add types again now, so maybe it's okay.

File: packages/gatsby/src/schema/infer/merge-inferred-composer.js [hunk 9]
```diff
@@ -0,0 +1,165 @@
+const _ = require(`lodash`)
```
Comment (wardpeet) [inline packages/gatsby/src/schema/infer/merge-inferred-composer.js]: it would be handy if you write some jsdoc comments above each function export to explain what it does as it's hard to follow if you have almost no knowledge of it.

File: packages/gatsby/src/schema/types/directives.js [hunk 10]
```diff
@@ -2,17 +2,18 @@ const {
...
+const { GraphQLJSON } = require(`graphql-compose`)
 
 const InferDirective = new GraphQLDirective({
   name: `infer`,
   description: `Infer fields for this type from nodes.`,
   locations: [DirectiveLocation.OBJECT],
   args: {
     noDefaultResolvers: {
-      type: new GraphQLNonNull(GraphQLBoolean),
-      default: false,
```
[the remark this answers is not in the document]
Comment (freiksenet) [reply packages/gatsby/src/schema/types/directives.js]: It's so we could distinguish between `null`, `true` and `false`.

File: packages/gatsby/src/schema/infer/merge-inferred-composer.js (hunk 9 above)
Comment (freiksenet) [reply packages/gatsby/src/schema/infer/merge-inferred-composer.js]: :+1: good idea

File: packages/gatsby/src/schema/infer/add-inferred-fields.js [hunk 11]
```diff
@@ -325,21 +268,22 @@ const getSimpleFieldConfig = ({
...
+        return {
+          type: `File`,
+          extensions: { addResolver: { type: `relativeFile` } },
+        }
       }
       return { type: `String` }
     case `object`:
       if (value instanceof Date) {
-        return dateResolver
+        return { type: `Date`, extensions: { addResolver: { type: `date` } } }
```
[the remark this answers is not in the document]
Comment (freiksenet) [reply packages/gatsby/src/schema/infer/add-inferred-fields.js]: done

File: packages/gatsby/src/schema/add-field-resolvers.js [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+  typeComposer,
+  fieldName,
+  options: { defaultFormat, defaultLocale },
+}) => {
+  const field = typeComposer.getField(fieldName)
+
+  let fieldConfig = {
+    resolve: dateResolver.resolve,
+  }
+  if (!field.args || _.isEmpty(field.args)) {
```
[the remark this answers is not in the document]
Comment (freiksenet) [reply packages/gatsby/src/schema/add-field-resolvers.js]: We don't want to override args or defaults if they are already defined somehow (by user in create types for example).

File: packages/gatsby/src/schema/types/__tests__/date.js [hunk 13]
```diff
@@ -423,12 +398,12 @@ describe(`dateResolver`, () => {
     expect(fields.invalidDate5.resolve).toBeUndefined()
     expect(fields.invalidDate6.resolve).toBeUndefined()
     expect(fields.invalidDate7.resolve).toBeUndefined()
-    expect(fields.invalidDate8.resolve).toBeUndefined()
+    expect(fields.invalidDate8).toBeUndefined()
```
[the remark this answers is not in the document]
Comment (freiksenet) [reply packages/gatsby/src/schema/types/__tests__/date.js]: The test was funky cause it didn't really test anything - it just assigned everything to Date. This actually tests that invalid dates don't get types. In addition, fields that are all null/undefined don't get graphql fields at all.

File: packages/gatsby/src/schema/types/directives.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (wardpeet) [reply packages/gatsby/src/schema/types/directives.js]: GraphQLNonNull make sense 🤦‍♂️

File: packages/gatsby/src/schema/add-field-resolvers.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (wardpeet) [reply packages/gatsby/src/schema/add-field-resolvers.js]: ah yeah mine merges and you don't want to so 👍

**Comment by KyleAMathews** [comment]
Why aren't we deprecating things now and then removing them in v3?

**Comment by freiksenet** [comment]
@KyleAMathews that's what we are doing, i phrased it awkwardly. Will fix.

**Comment by stefanprobst** [comment]
I have tried to simplify some things a bit in  #13557, feel free to cherry-pick.
* some changes to third-party type merging were necessary because of https://github.com/graphql-compose/graphql-compose/pull/183 upstream, but I was not sure of the reason for the rest of the changes? Maybe needs a test if there is?
* don't use temporary composers in inference. this needed changing as the upstream change in `getFieldTC` broke things here - and it seems temp composers will be deprecated in a future version. anyway we don't need them anymore.

**Comment by freiksenet** [comment]
@stefanprobst I think some of the changes to third party types could have been caused by some regression in graphql-compose. If it works without them (as in passes tests), then it's fine.

**Comment by stefanprobst** [comment]
Last time I bring up API naming, I promise:
* I'd prefer `dateformat` over `date` (because it is about formatting), and keep the options in sync with args, i.e. `formatString` and `locale`
* `fileByPath` or `fileByRelativePath` sounds more descriptive than `relativeFile` to me

**Comment by CanRau** [comment]
This PR is not about disabling "..There are conflicting field types in your data.." if custom types have been defined right? I came here via https://github.com/gatsbyjs/gatsby/issues/12272#issuecomment-480184188 which I think has already the answer, just wanted to understand and bring this confusing warning up (I believe you know about it anyway?!). Took me a while and some googling to realize it's a "false friend" and to be ignored, thougt my types had problems 😅

**Comment by stefanprobst** [comment]
@CanRau If you define a type with the `@dontInfer` directive (or with `extensions: { infer: false }` if you are using Gatsby Type Builders), this PR will get rid of the warning message, since type inference will be totally skipped in that case. This of course requires that you define all field types explicitly and don't rely on type inference. Example:
```graphql
type TestType implements Node @dontInfer {
  foo: Int
  bar: Int
}
```
In cases where type inference runs, I agree that the warning message is a bit confusing currently. What do you think about the changes in #13693?

**Comment by CanRau** [comment]
Uuuh okay, 🙏 thanks for clarifying, that makes sense 💡 thought you guys were talking about disabling it altogether 🙄 so only for the custom typed one..I like it 👍 sounds super handy 👏 

> In cases where type inference runs, I agree that the warning message is a bit confusing currently. What do you think about the changes in #13693?

A little verbose 😅 yet waay better than currently very confusing situation 👏

**Comment by freiksenet** [comment]
Published `2.4.0-alpha.3`.

File: docs/blog/2019-04-23-improvements-to-schema-customization/index.md [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+
+## `addResolver`
+
+Add resolver and resolver options (such as arguments) to the given field.
+
+```graphql
+type MyType @infer {
+  date: Date @addResolver(type: "dateformat", options: { formatString: "DD MMM", locale: "fi" })
+  image: File @addResolver(type: "fileByRelativePath")
+  authorByEmail: Author @addResolver(type: "link", { by: "email" })
```
Comment (stefanprobst) [inline docs/blog/2019-04-23-improvements-to-schema-customization/index.md]:
```suggestion
  authorByEmail: Author @addResolver(type: "link", options: { by: "email" })
```

**Comment by freiksenet** [comment]
@stefanprobst Could you do last review?

@DSchau Could you do blog post review?

**Comment by freiksenet** [comment]
Published `gatsby@2.5.0-rc.1`

File: docs/blog/2019-05-17-improvements-to-schema-customization/index.md [hunk 15]
```diff
@@ -0,0 +1,107 @@
+---
+title: Improvements to Schema Customization API - Available in Gatsby 2.4.0
```
Comment (m-allanson) [inline docs/blog/2019-05-17-improvements-to-schema-customization/index.md]: These should all be changed from 2.4.0 to 2.5.0?

```suggestion
title: Improvements to Schema Customization API - Available in Gatsby 2.5.0
```

File: docs/blog/2019-05-17-improvements-to-schema-customization/index.md [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+## Summary
+
+After about a month of testing schema customization both here and in pre-release we determined a couple of issues. We set up to do this because we wanted to remove uncertainty in people's schemas, when the data changes. However, the original design allowed some uncertainties anyway. In addition, we have made inferrence a more heavy process, trading performance for consistency and didn't really provide a way to opt out of it completely. To summarize:
+
+- Resolvers and arguments of fields like Date and File was determined by inferred data
+- There was no easy way to use arguments/resolvers to override the above
+- Inferrence was run even when `@dontInfer` flag was on
+- There was no way to control inference outside of SDL, eg in Type Builders
+
+Therefore we have some changes to the way we do inferrence. In addition, we are deprecating some of the features introduced is 2.2.0 and will remove them in Gatsby 3.
```
Comment (m-allanson) [inline docs/blog/2019-05-17-improvements-to-schema-customization/index.md]:
```suggestion
Therefore we have some changes to the way we do inferrence. In addition, we are deprecating some of the features introduced in 2.2.0 and will remove them in Gatsby 3.
```

File: docs/blog/2019-05-17-improvements-to-schema-customization/index.md [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+  - graphql
+---
+
+Today we are releasing further improvements to the schema customization that [we've released in version 2.2.0](/blog/2019-03-18-releasing-new-schema-customization). You can use them with Gatsby 2.4.0.
+
+It is now possible to indicate to Gatsby, that you want to add a resolver to an explicitly defined fields. Use extensions like `@link` and `@dateformat` to add default arguments or/and resolvers to fields. In addition, when `@dontInfer` is set, Gatsby will no longer run inference for marked type, allowing one to improve performance for large data sets.
+
+## Summary
+
+After about a month of testing schema customization both here and in pre-release we determined a couple of issues. We set up to do this because we wanted to remove uncertainty in people's schemas, when the data changes. However, the original design allowed some uncertainties anyway. In addition, we have made inferrence a more heavy process, trading performance for consistency and didn't really provide a way to opt out of it completely. To summarize:
```
Comment (m-allanson) [inline docs/blog/2019-05-17-improvements-to-schema-customization/index.md]: I took a go at expanding this to make it clearer that it's talking about issues with the 2.2.0 release of schema customization...

```suggestion
After about a month of testing schema customization both here and in pre-release we determined a couple of issues. The original aim of our schema customisation work was to remove uncertainty in people's schemas when their data changes. 

However, the original design allowed some uncertainties anyway. In addition, it made inference a more heavy process, trading performance for consistency without providing a way to opt out completely. To summarize, the schema customization work released in Gatsby 2.2.0 had the following issues:
```

File: docs/blog/2019-05-17-improvements-to-schema-customization/index.md [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+
+- Resolvers and arguments of fields like Date and File was determined by inferred data
+- There was no easy way to use arguments/resolvers to override the above
+- Inferrence was run even when `@dontInfer` flag was on
+- There was no way to control inference outside of SDL, eg in Type Builders
+
+Therefore we have some changes to the way we do inferrence. In addition, we are deprecating some of the features introduced is 2.2.0 and will remove them in Gatsby 3.
+
```
Comment (m-allanson) [inline docs/blog/2019-05-17-improvements-to-schema-customization/index.md]: Could add a heading here, something like "Updates in Gatsby 2.5.0"? Saying something about "these are the problems we identified, here are the updates we've created to fix them"

This would make it really clear where you've stopped talking about 2.2.0 problems and started talking about 2.5.0 updates.

File: docs/blog/2019-05-17-improvements-to-schema-customization/index.md [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+## noDefaultResolvers and inferrence modes
+
+First of all, we are deprecating `noDefaultResolvers`. It was an argument of `infer` and `dontInfer`. We feel it was confusing and in some cases it didn't even actually add resolvers :). We will support `noDefaultResolvers` until version 3, after which `@infer` behaviour (see below) will become a default and `noDefaultResolvers` will be removed.
```
Comment (m-allanson) [inline docs/blog/2019-05-17-improvements-to-schema-customization/index.md]: Some before and after code examples might be useful in these sections.

File: docs/blog/2019-05-17-improvements-to-schema-customization/index.md [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+      },
+    },
+  },
+})
+```
+
+## Conclusions
+
+With these improvements we hope we'll solve most of the issues that people are having with new schema customization. We are working on further improvements, like allowing users and plugins to define their own extensions (see [PR #13738](https://github.com/gatsbyjs/gatsby/pull/13738)).
+
```
Comment (m-allanson) [inline docs/blog/2019-05-17-improvements-to-schema-customization/index.md]: Could specifically ask for feedback and point out the umbrella issue link here?

File: packages/gatsby/src/redux/actions.js [hunk 21]
```diff
@@ -1294,9 +1314,9 @@ import type GatsbyGraphQLType from "../schema/types/type-builders"
  *     Author information
  *     """
  *     # Does not include automatically inferred fields
- *     type AuthorJson implements Node @dontInfer(noFieldResolvers: true) {
+ *     type AuthorJson implements Node @dontInfer {
  *       name: String!
- *       birthday: Date! # no default resolvers for Date formatting added
+ *       birthday: Date! @date(locale: "ru")
```
Comment (stefanprobst) [inline packages/gatsby/src/redux/actions.js]:
```suggestion
 *       birthday: Date! @dateformat(locale: "ru")
```

File: packages/gatsby/src/redux/actions.js [hunk 22]
```diff
@@ -1322,12 +1345,40 @@ import type GatsbyGraphQLType from "../schema/types/type-builders"
...
- *     })
+ *     }),
+ *     schema.buildObjectType({
+ *       name: 'AuthorJson',
+ *       fields: {
+ *         name: 'String!'
+ *         birthday: {
+ *           type: 'Date!'
+ *           extensions: {
+ *             date: {
```
Comment (stefanprobst) [inline packages/gatsby/src/redux/actions.js]:
```suggestion
 *             dateformat: {
```

File: docs/blog/2019-05-17-improvements-to-schema-customization/index.md (hunk 18 above)
Comment (m-allanson) [reply docs/blog/2019-05-17-improvements-to-schema-customization/index.md]: Done in `f338bdd` 👍

File: docs/blog/2019-05-17-improvements-to-schema-customization/index.md (hunk 20 above)
Comment (m-allanson) [reply docs/blog/2019-05-17-improvements-to-schema-customization/index.md]: Done in `f338bdd` 👍

File: docs/blog/2019-05-17-improvements-to-schema-customization/index.md (hunk 19 above)
Comment (m-allanson) [reply docs/blog/2019-05-17-improvements-to-schema-customization/index.md]: Done in `f338bdd` 👍
